### PR TITLE
Remove 'synchronized (this)' blocking for writing network state file in updateNode/addNode/removeNode

### PR DESCRIPTION
**Pull request by puzzle-star**
In updateNode / addNode / removeNode, access to nodeListeners list does not need to be synchronized as, the list is unmodifiable (and fully replaced by another one when listeners are added / removed)

Serializing afterwards needs to be synchronized, but it can be done in a separate thread, so this is called now through a separate thread from the executorService.

### CODE REVIEW

File: com.zsmartsystems.zigbee/src/main/java/com/zsmartsystems/zigbee/ZigBeeNetworkManager.java [hunk 1]
```diff
@@ -1295,22 +1295,27 @@ public void removeNode(final ZigBeeNode node) {
...
-            node.shutdown();
+        node.shutdown();
 
-            if (networkStateSerializer != null) {
-                networkStateSerializer.serialize(this);
+        executorService.execute(new Runnable() {
+            @Override
+            public void run() {
+                synchronized (ZigBeeNetworkManager.this) {
+                    if (networkStateSerializer != null) {
```
Comment (cdjackson) [inline com.zsmartsystems.zigbee/src/main/java/com/zsmartsystems/zigbee/ZigBeeNetworkManager.java]: Wouldn't it be better to put this check outside of the thread?

File: com.zsmartsystems.zigbee/src/main/java/com/zsmartsystems/zigbee/ZigBeeNetworkManager.java [hunk 2]
```diff
@@ -1335,24 +1340,29 @@ public void addNode(final ZigBeeNode node) {
...
+            });
+        }
 
-            if (networkStateSerializer != null) {
-                networkStateSerializer.serialize(this);
+        executorService.execute(new Runnable() {
+            @Override
+            public void run() {
+                synchronized (ZigBeeNetworkManager.this) {
+                    if (networkStateSerializer != null) {
```
Comment (cdjackson) [inline com.zsmartsystems.zigbee/src/main/java/com/zsmartsystems/zigbee/ZigBeeNetworkManager.java]: Wouldn't it be better to put this check outside of the thread?

File: com.zsmartsystems.zigbee/src/main/java/com/zsmartsystems/zigbee/ZigBeeNetworkManager.java [hunk 3]
```diff
@@ -1389,24 +1399,29 @@ public void updateNode(final ZigBeeNode node) {
...
```
Comment (cdjackson) [inline com.zsmartsystems.zigbee/src/main/java/com/zsmartsystems/zigbee/ZigBeeNetworkManager.java]: Wouldn't it be better to put this check outside of the thread?

File: com.zsmartsystems.zigbee/src/main/java/com/zsmartsystems/zigbee/ZigBeeNetworkManager.java (hunk 1 above)
Comment (puzzle-star) [reply com.zsmartsystems.zigbee/src/main/java/com/zsmartsystems/zigbee/ZigBeeNetworkManager.java]: If it were outside, networkStateSerializer could be set to null by another thread between the check and calling the scheduled method, leading to a NPE.
Setting networkStateSerializer is done synchronized on (this).

Comment (puzzle-star) [reply com.zsmartsystems.zigbee/src/main/java/com/zsmartsystems/zigbee/ZigBeeNetworkManager.java]: If it were outside, networkStateSerializer could be set to null by another thread between the check and calling the scheduled method, leading to a NPE.
Setting networkStateSerializer is done synchronized on (this).

Comment (puzzle-star) [reply com.zsmartsystems.zigbee/src/main/java/com/zsmartsystems/zigbee/ZigBeeNetworkManager.java]: If it were outside, networkStateSerializer could be set to null by another thread between the check and calling the scheduled method, leading to a NPE.
Setting networkStateSerializer is done synchronized on (this).

File: com.zsmartsystems.zigbee/src/main/java/com/zsmartsystems/zigbee/ZigBeeNetworkManager.java (hunk 2 above)
Comment (puzzle-star) [reply com.zsmartsystems.zigbee/src/main/java/com/zsmartsystems/zigbee/ZigBeeNetworkManager.java]: Same as above ...

File: com.zsmartsystems.zigbee/src/main/java/com/zsmartsystems/zigbee/ZigBeeNetworkManager.java (hunk 3 above)
Comment (puzzle-star) [reply com.zsmartsystems.zigbee/src/main/java/com/zsmartsystems/zigbee/ZigBeeNetworkManager.java]: Same as above...

**Comment by cdjackson** [comment]
Fair enogh - do you plan to do an update?

**Comment by puzzle-star** [comment]
I am not at home now. I can move the duplication to a separate method later (i.e triggerNetworkSerialization), but it will have to be changed for the datastore, and maybe the three calls will be different then, do maybe it is not worth it...

Let me know what you prefer

Pedro

**Comment by cdjackson** [comment]
Ok, let’s leave it as is. I hope to release 1.2 this month anyway.

**Pull request merged**